compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -640,6 +640,24 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
         NoType
       }
     }
+
+    // The intersection between a singleton type s and another type t is
+    // always empty when s is not a subtype of t. See patmat/i3573.scala
+    val hasEmptyIntersections = new ExistsAccumulator({
+      case AndType(s: SingletonType, t) =>
+        !(s <:< t)
+      case AndType(t, s: SingletonType) =>
+        !(s <:< t)
+      case x =>
+        false
+    })
+
+    if (hasEmptyIntersections(false, result)) {
+      debug.println(s"hasEmptyIntersections($protoTp1) = true")
+      NoType
+    } else {
+      result
+    }
   }
```

Comment:
In the check, there's a facility named `implementability` from a contributor which does more or less the same thing. What do you think we can generalise/adapt `implementability` to handle the case?

In particular, I think it's good to rename it to `def inhabited(tp: Type): Boolean`, and it should be able to handle the current case as well. WDYT?

Reply:
I think `implementability` computes if a type can be implemented by other types (singleton types can't!), while `inhabited` would need to check if there could actually exists an instance of that type, so I'm not sure two logics can be merged. @AleksanderBG what's your opinion on that?

Reply:
I think `implementability` should be refactored to have the semantics of `inhabited`. According to the core algorithm of exhaustivity check, it should only depend on the semantics of `def inhabited(tp: Type): Boolean`.

Reply:
@OlivierBlanvillain I believe @liufengyun is correct - the function can be safely changed to check if a type is inhabited instead. This won't change the result for and-types, which is the only result `intersectUnrelatedAtomicTypes` should care about. I considered a similar refactoring myself when working on #3454. With the benefit of hindsight, type inhabitation is a far more generic notion and I should've tried to implement it from the beginning.